src/librustdoc/clean/types.rs

```diff
@@ -1304,6 +1302,49 @@ crate enum TypeKind {
     TraitAlias,
 }
 
+impl<'a> From<&'a hir::def::DefKind> for TypeKind {
+    fn from(other: &hir::def::DefKind) -> Self {
+        Self::from(*other)
+    }
+}
+
+impl From<hir::def::DefKind> for TypeKind {
+    fn from(other: hir::def::DefKind) -> Self {
+        match other {
+            hir::def::DefKind::Enum => Self::Enum,
+            hir::def::DefKind::Fn => Self::Function,
+            hir::def::DefKind::Mod => Self::Module,
+            hir::def::DefKind::Const => Self::Const,
+            hir::def::DefKind::Static => Self::Static,
+            hir::def::DefKind::Struct => Self::Struct,
+            hir::def::DefKind::Union => Self::Union,
+            hir::def::DefKind::Trait => Self::Trait,
+            hir::def::DefKind::TyAlias => Self::Typedef,
+            hir::def::DefKind::TraitAlias => Self::TraitAlias,
+            hir::def::DefKind::Macro(_) => Self::Macro,
+            hir::def::DefKind::ForeignTy
+            | hir::def::DefKind::Variant
+            | hir::def::DefKind::AssocTy
+            | hir::def::DefKind::TyParam
+            | hir::def::DefKind::ConstParam
+            | hir::def::DefKind::Ctor(..)
+            | hir::def::DefKind::AssocFn
+            | hir::def::DefKind::AssocConst
+            | hir::def::DefKind::ExternCrate
+            | hir::def::DefKind::Use
+            | hir::def::DefKind::ForeignMod
+            | hir::def::DefKind::AnonConst
+            | hir::def::DefKind::OpaqueTy
+            | hir::def::DefKind::Field
+            | hir::def::DefKind::LifetimeParam
+            | hir::def::DefKind::GlobalAsm
+            | hir::def::DefKind::Impl
+            | hir::def::DefKind::Closure
+            | hir::def::DefKind::Generator => Self::Foreign,
```

Comment:
This behavior isn't from this PR, but why do we treat all of these things as having a "foreign type"? Some of them don't have a type (e.g. `use`), so it seems like we should return an optional value and `None` in these cases instead.

And for other things, like `OpaqueTy` (for that one I don't even understand why it has a `DefKind`, given that it's not "defined") and `TyParam`, it seems we should create a new variant `TypeKind::Other` for things we don't explicitly handle.

Reply:
I'll add a comment about it but it's from the search index perspective: if a type isn't the search index, then we consider it as a "foreign" one to ignore it.